src/coreclr/jit/gentree.cpp

```diff
@@ -1634,262 +1634,68 @@ bool GenTree::Compare(GenTree* op1, GenTree* op2, bool swapOK)
     return false;
 }
 
-/*****************************************************************************
- *
- *  Returns non-zero if the given tree contains a use of a local #lclNum.
- */
-
-// TODO-List-Cleanup: rewrite with a general visitor.
-bool Compiler::gtHasRef(GenTree* tree, ssize_t lclNum, bool defOnly)
+//------------------------------------------------------------------------
+// gtHasRef: Find out whether the given tree contains a local/field.
+//
+// Arguments:
+//    tree    - tree to find the local in
+//    lclNum  - the local's number, *or* the handle for the field
+//
+// Return Value:
+//    Whether "tree" has any LCL_VAR nodes that refer to the local,
+//    LHS or RHS, or FIELD nodes with the specified handle.
+//
+// Notes:
+//    Does not pay attention to local address nodes or LCL_FLD nodes.
+//
+/* static */ bool Compiler::gtHasRef(GenTree* tree, ssize_t lclNum)
 {
-    genTreeOps oper;
-    unsigned   kind;
-
-AGAIN:
-
-    assert(tree);
-
-    oper = tree->OperGet();
-    kind = tree->OperKind();
-
-    /* Is this a constant node? */
-
-    if (kind & GTK_CONST)
+    if (tree == nullptr)
     {
         return false;
     }
 
-    /* Is this a leaf node? */
-
-    if (kind & GTK_LEAF)
+    if (tree->OperIsLeaf())
     {
-        if (oper == GT_LCL_VAR)
+        if (tree->OperIs(GT_LCL_VAR) && (tree->AsLclVarCommon()->GetLclNum() == (unsigned)lclNum))
         {
-            if (tree->AsLclVarCommon()->GetLclNum() == (unsigned)lclNum)
-            {
-                if (!defOnly)
-                {
-                    return true;
-                }
-            }
+            return true;
         }
-        else if (oper == GT_RET_EXPR)
+        if (tree->OperIs(GT_RET_EXPR))
         {
-            return gtHasRef(tree->AsRetExpr()->gtInlineCandidate, lclNum, defOnly);
+            return gtHasRef(tree->AsRetExpr()->gtInlineCandidate, lclNum);
         }
 
         return false;
     }
 
-    /* Is it a 'simple' unary/binary operator? */
-
-    if (kind & GTK_SMPOP)
+    if (tree->OperIsUnary())
     {
-        // Code in importation (see CEE_STFLD in impImportBlockCode), when
-        // spilling, can pass us "lclNum" that is actually a field handle...
-        if (tree->OperIs(GT_FIELD) && (lclNum == (ssize_t)tree->AsField()->gtFldHnd) && !defOnly)
+        if (tree->OperIs(GT_FIELD) && (lclNum == (ssize_t)tree->AsField()->gtFldHnd))
```

Comment:
I realize this is pre-existing code, but it creates a really weird function signature. Do we really know that the values of gtFldHnd are guaranteed to be distinct from local numbers? Also, based on the previous comment that was here (which I would leave if this code is still required), "see CEE_STFLD in impImportBlockCode" -- I don't see any case in the JIT, including that location, that appears to pass a field handle to this function. Maybe this code is dead code? Can we eliminate this case and assume it's a local? Even if we can't, it seems like we should have a completely separate function that takes a field handle, and not overload this one function argument for both purposes.

Reply:
The code is, unfortunately, not dead. The "STFLD" case that it is (was) talking about is here:

https://github.com/dotnet/runtime/blob/e43688a3fe6957f669ac10bc5413e285773732c3/src/coreclr/jit/importer.cpp#L15953-L15982

I will bring [the comment](https://github.com/dotnet/runtime/blame/e43688a3fe6957f669ac10bc5413e285773732c3/src/coreclr/jit/gentree.cpp#L1688-L1689) back.

> Do we really know that the values of `gtFldHnd` are guaranteed to be distinct from local numbers?

No. The Jit depends on the invariant that the field handles do not have the two lower bits set, and Crossgen2 makes this work with some explicit code, but I don't think it takes the sort of punning this method is doing into account. With the CLR VM we're "safe" because the handles are always pointers there.

In any case, I think it is not correctness issue if we "overestimate" what we see in this method.

> it seems like we should have a completely separate function that takes a field handle, and not overload this one function argument for both purposes.

I very much agree. I am currently working on helpers that would allow for writing this sort of code with minimal boilerplate (and on-par performance), and this method will be split into two as part of that work. Let me know if you think it better to do that as part of this PR though.

Reply:
Great, thanks for the explanation; I didn't find that case when I looked.

> Let me know if you think it better to do that as part of this PR though.

No, I'll merge this as-is.

Thanks for the contribution!